xmtp_api_http/src/http_stream.rs

```diff
@@ -0,0 +1,314 @@
+    }
+}
+
+pin_project! {
+    struct HttpPostStream<'a, R> {
+        #[pin] http: StreamWrapper<'a, Result<bytes::Bytes, reqwest::Error>>,
+        remaining: Vec<u8>,
+        _marker: PhantomData<&'a R>,
+    }
+}
+
+impl<R> Stream for HttpPostStream<'_, R>
+where
+    for<'de> R: Send + Deserialize<'de>,
+{
+    type Item = Result<R, Error>;
+
+    fn poll_next(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> Poll<Option<Self::Item>> {
+        use Poll::*;
+        let mut this = self.as_mut().project();
+        let item = ready!(this.http.as_mut().poll_next(cx));
+        match item {
+            Some(bytes) => {
+                let bytes = bytes
+                    .inspect_err(|e| tracing::error!("Error in http stream to grpc gateway {e}"))
+                    .map_err(|_| Error::new(ErrorKind::SubscribeError))?;
+                let item = Self::on_bytes(bytes, this.remaining)?.pop();
+                if let Some(item) = item {
+                    Ready(Some(Ok(item)))
+                } else {
+                    self.poll_next(cx)
+                }
+            }
+            None => Ready(None),
+        }
+    }
+}
+
+impl<'a, R> HttpPostStream<'a, R>
+where
+    R: Send + 'static,
+{
+    pub fn new(establish: StreamWrapper<'a, Result<bytes::Bytes, reqwest::Error>>) -> Self {
+        Self {
+            http: establish,
+            remaining: Vec::new(),
+            _marker: PhantomData,
+        }
+    }
+}
+
+impl<R> HttpPostStream<'_, R>
+where
+    for<'de> R: Deserialize<'de> + DeserializeOwned + Send,
+{
+    fn on_bytes(bytes: bytes::Bytes, remaining: &mut Vec<u8>) -> Result<Vec<R>, Error> {
+        let bytes = &[remaining.as_ref(), bytes.as_ref()].concat();
+        let de = Deserializer::from_slice(bytes);
+        let mut deser_stream = de.into_iter::<GrpcResponse<R>>();
+        let mut items = Vec::new();
+        loop {
+            let item = deser_stream.next();
+            if item.is_none() {
+                break;
+            }
+            match item.expect("checked for none;") {
+                Ok(GrpcResponse::Ok(response)) => items.push(response),
+                Ok(GrpcResponse::SubscriptionItem(item)) => items.push(item.result),
+                Ok(GrpcResponse::Err(e)) => {
+                    return Err(Error::new(ErrorKind::MlsError).with(e.message));
+                }
+                Err(e) => {
+                    if e.is_eof() {
+                        *remaining = (&**bytes)[deser_stream.byte_offset()..].to_vec();
+                        break;
+                    } else {
+                        return Err(Error::new(ErrorKind::MlsError).with(e.to_string()));
+                    }
+                }
+                Ok(GrpcResponse::Empty {}) => continue,
+            }
+        }
+
+        if items.len() > 1 {
+            tracing::warn!("more than one item deserialized from http stream");
```

Comment:
Why is this worth warning?

Reply:
can remove it, but I guess it's defensive. I'm not sure if grpc http api would ever send two items in a response. I don't think it ever should, but if it does that's what the warn would be for. If it does send two items we would probably miss messages